text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+reference, we know that it *can* move self, or even a field of self.
+
+Moving *fields* of self is a particular problem because of how it interacts
+with pin *projections* - that is, taking a Pin reference to type and
+"projecting" to a pinned reference to a field of that type. If you move out of
+the fields of a struct in the struct's destructor, pin projecting to that field
+would be unsound, because it would have been moved after it was pinned and
+before its destructor ran.
+
+This applies equally well to `poll_drop_ready`. If `poll_drop_ready` took self
+by `Pin`, in order to generate drop glue it would need to pin project to each
+field of the type. The order of operations would be:
+
+1. `poll_drop_ready` runs, pin projecting to each field of the type.
+2. The type's destructor runs, potentially moving each field.
+3. Each field's destructor runs, after the type's destructor has potentially
+moved them.
+
+Because of this sequence, the in projection in `poll_drop_ready` would be
+unsound. Therefore, `poll_drop_ready` cannot take self by Pin. Here is a bit of
+code demonstrating what I mean:
+
+```rust
+// If `field` implements `poll_drop_ready` and `poll_drop_ready` used `Pin`,
+// `field` would be witnessed as pinned before calling drop on `Foo`.
+struct Foo<T: Default> {
+    field: T
+}
+
+impl<T: Default> Drop for Foo<T> {
+    fn drop(&mut self) {
+        // Then, in the destructor for `Foo`, we can move `field` and re-pin it
+        // at a different location, violating the guarantees of `Pin`.
+
+        let moved_field = mem::replace(&mut self.field, T::default());
+        let pinned: Box<Pin<T>> = Box::pin(moved_field);
+    }
+}
+```
+
+However, its worth noting that a user can locally reason about whether private
+fields can be moved or accessed again, based on the APIs exposed by a type,
+and, if necessary, reconstruct a pin of those fields. [The pin documentation
+contains relevant information.][pin-docs] In other words, if your drop
+implementation is implemented correctly, you can construct a `Pin` of self
+inside your `poll_drop_ready`.
+
+## Drop order
+
+The order of calls to `poll_drop_ready` can be inferred from the rest of this
+document, but it's good to specify them explicitly.
+
+Between variables, calls to `poll_drop_ready` will occur in reverse order of
+the variable's introduction, immediately prior to the calls to that variable's
+destructor.
+
+Between fields of a type, calls to `poll_drop_ready` will occur in the textual
+order of that type's declaration, with the call for the type itself occuring
+first. This is similar to the order for calls to drop. *However*, these calls
+will occur in a loop, at the level of the type being dropped, until all calls
+to `poll_drop_ready` for that value have returned `Ready`, so they will be
+"interleaved" and concurrent in practice. The program will *not* wait for each
+field to return ready before beginning to process the subsequent field.
```

Comment:
Consider referencing https://github.com/rust-lang/rfcs/blob/master/text/1857-stabilize-drop-order.md here.

Also consider mentioning how built-in structures like `Vec` ought to behave and if there are or aren’t any special considerations in presence of panics, especially during structure construction.